### PR TITLE
 [crypto/ec] for ECC parameters with NULL or zero cofactor, compute it 

### PR DESCRIPTION
**Pull request by bbbrumley**
#6648 was a very quick fix to solve #6637. The `cofactor` argument to `EC_GROUP_set_generator` is optional, and SCA mitigations for ECC currently use it. So the library currently falls back to very old SCA-vulnerable code if the cofactor is not present.

This PR allows `EC_GROUP_set_generator` to compute the cofactor for all curves of cryptographic interest. Steering scalar multiplication to more SCA-robust code.

Affects all non-EOL branches.

We coordinated this disclosure with `openssl-security`.

### CODE REVIEW

File: crypto/ec/ec_lib.c [hunk 1]
```diff
@@ -274,6 +274,68 @@ int EC_METHOD_get_field_type(const EC_METHOD *meth)
...
+
+    /* compute h = \lfloor (q + 1)/n \rceil = \lfloor (q + 1 + n/2)/n \rfloor */
+    if (!BN_rshift1(group->cofactor, group->order) /* n/2 */
+        || !BN_add(group->cofactor, group->cofactor, q) /* q + n/2 */
+        /* q + 1 + n/2 */
+        || !BN_add(group->cofactor, group->cofactor, BN_value_one())
+        /* (q + 1 + n/2)/n */
+        || !BN_div(group->cofactor, NULL, group->cofactor, group->order, ctx))
+        goto err;
+    ret = 1;
```
Comment (romen) [inline crypto/ec/ec_lib.c]: If any of the above fails, should we BN_zero() the cofactor before returning 0?

File: crypto/ec/ec_lib.c [hunk 2]
```diff
@@ -274,6 +274,68 @@ int EC_METHOD_get_field_type(const EC_METHOD *meth)
...
+ */
+static int ec_guess_cofactor(EC_GROUP *group) {
+    int ret = 0;
+    BN_CTX *ctx = NULL;
+    BIGNUM *q = NULL;
+
+    if (group->field == NULL || group->order == NULL
+        || BN_is_zero(group->field) || BN_is_zero(group->order)
+        || BN_is_negative(group->field) || BN_is_negative(group->order)
+        || BN_num_bits(group->order) <= (BN_num_bits(group->field) + 1)/2 + 3) {
```
Comment (mattcaswell) [inline crypto/ec/ec_lib.c]: Nit: add spaces around "/"

File: crypto/ec/ec_lib.c [hunk 3]
```diff
@@ -274,6 +274,68 @@ int EC_METHOD_get_field_type(const EC_METHOD *meth)
...
+        /*-
+         * If the cofactor is too large, we cannot guess it. Default to 0.
+         * The RHS of above is a strict overestimate of lg(4 * sqrt(q))
+         */
+        BN_zero(group->cofactor);
+        return 1;
+    }
+
+    if ((ctx = BN_CTX_new_ex(group->libctx)) == NULL)
```
Comment (mattcaswell) [inline crypto/ec/ec_lib.c]: We're going to need a different version of this PR for all branches except master - `BN_CTX_new_ex` doesn't exist in those branches.

File: crypto/ec/ec_lib.c [hunk 4]
```diff
@@ -297,12 +359,16 @@ int EC_GROUP_set_generator(EC_GROUP *group, const EC_POINT *generator,
         BN_zero(group->order);
     }
 
-    /* The cofactor is an optional field, so it should be able to be NULL. */
-    if (cofactor != NULL) {
+    /*-
+     * Unfortunately the cofactor is an optional field. So it might be NULL.
+     * Either take the provided positive cofactor, or try to compute it.
+     */
+    if (cofactor != NULL && !BN_is_zero(cofactor) && !BN_is_negative(cofactor)) {
```
Comment (mattcaswell) [inline crypto/ec/ec_lib.c]: A negative cofactor just sounds like an error.

File: crypto/ec/ec_lib.c [hunk 5]
```diff
@@ -297,12 +359,16 @@ int EC_GROUP_set_generator(EC_GROUP *group, const EC_POINT *generator,
...
         if (!BN_copy(group->cofactor, cofactor))
             return 0;
-    } else {
-        BN_zero(group->cofactor);
+    }
+    else if (!ec_guess_cofactor(group)) {
```
Comment (mattcaswell) [inline crypto/ec/ec_lib.c]: Nit: Put else on same line as }

**Comment by mattcaswell** [comment]
Note: I've not checked the maths

File: crypto/ec/ec_lib.c (hunk 2 above)
Comment (bbbrumley) [reply crypto/ec/ec_lib.c]: Fixed in c628486f8c59450a6e6722f658dc6ba40743da19 force push

File: crypto/ec/ec_lib.c (hunk 4 above)
Comment (bbbrumley) [reply crypto/ec/ec_lib.c]: Fixed in c628486f8c59450a6e6722f658dc6ba40743da19 force push

File: crypto/ec/ec_lib.c (hunk 5 above)
Comment (bbbrumley) [reply crypto/ec/ec_lib.c]: Fixed in c628486f8c59450a6e6722f658dc6ba40743da19 force push

File: crypto/ec/ec_lib.c (hunk 1 above)
Comment (bbbrumley) [reply crypto/ec/ec_lib.c]: Fixed in c628486f8c59450a6e6722f658dc6ba40743da19 force push

File: test/ectest.c [hunk 6]
```diff
@@ -1679,9 +1679,8 @@ static int check_named_curve_test(int id)
                                              group_cofactor))
         || !TEST_int_eq(EC_GROUP_check_named_curve(gtest, 0, NULL), 0)
         /* The order is not an optional field, so this should fail */
-        || !TEST_true(EC_GROUP_set_generator(gtest, group_gen, NULL,
+        || TEST_true(EC_GROUP_set_generator(gtest, group_gen, NULL,
```
Comment (romen) [inline test/ectest.c]: nit: spacing of following line is off by 1

**Comment by bbbrumley** [comment]
Based on the feedback, latest force push addresses the comments but also:
* throws errors if `group->field` isn't properly set
* throws errors if `order` argument is not sane
* throws errors if `cofactor` argument is not sane (outside `NULL` / zero)
* negative tests for the above

File: crypto/ec/ec_lib.c [hunk 7]
```diff
@@ -282,6 +343,34 @@ int EC_GROUP_set_generator(EC_GROUP *group, const EC_POINT *generator,
...
+    /*-
+     * - require order >= 1
+     * - enforce upper bound due to Hasse thm: order can be no more than one bit
+     *   longer than field cardinality
+     */
+    if (order == NULL || BN_is_zero(order) || BN_is_negative(order)
+        || BN_num_bits(order) > BN_num_bits(group->field) + 1) {
+        ECerr(EC_F_EC_GROUP_SET_GENERATOR, EC_R_INVALID_GROUP_ORDER);
+        return 0;
+    }
```
Comment (romen) [inline crypto/ec/ec_lib.c]: We discussed this f2f with @bbbrumley and I support this change: we are basically repeating here all the checks that the library is already doing during ASN.1 parsing for the order field before calling this function.

The previous behavior of allowing to programmatically set the `order` to zero by passing `NULL` is just silly: we can't do anything with an `EC_GROUP` object with `order == 0` anyway, and it was possible to create it only programmatically not parsing a file.

This is something an other reviewer might want to evaluate closely, but I totally support this change also for backporting to the other active branches.

File: test/ectest.c (hunk 6 above)
Comment (romen) [reply test/ectest.c]: This unit test is `master` only currently so it is fine to modify it and make sure a failure is returned if the mandatory `order` argument is not valid (`NULL`).

We lose a little bit of coverage on `EC_GROUP_check_named_curve()` but now, rationalizing the behaviour of `EC_GROUP_set_generator()`, the only way the former can process an `EC_GROUP` that has `order == 0` is if the `EC_GROUP` has been created from scratch and the generator is never set before calling `EC_GROUP_check_named_curve()`.

This case basically means that the `EC_GROUP` has not been properly initialized before calling `EC_GROUP_check_named_curve()` but I don't think we test any other function for this corner case condition.

@slontis do you think it's worth it to add a separate check to test `EC_GROUP_check_named_curve()`'s behaviour in this corner case?

Comment (bbbrumley) [reply test/ectest.c]: > nit: spacing of following line is off by 1

fixed in 23540664ac7e2361fe3bb191a28a21c0cd7890f8

File: crypto/ec/ec_lib.c (hunk 7 above)
Comment (mattcaswell) [reply crypto/ec/ec_lib.c]: Hmm. That would be a technical compatibility break. It's not unreasonable to expect that some applications may actually set these values in multiple separate calls to this function, i.e. just because order is NULL on one call, it doesn't mean it will still be NULL by the time the group is actually used.

Comment (bbbrumley) [reply crypto/ec/ec_lib.c]: > just because order is NULL on one call, it doesn't mean it will still be NULL by the time the group is actually used

I understand. But how else would `order` be set?

Comment (mattcaswell) [reply crypto/ec/ec_lib.c]: So consider this code:

    EC_GROUP_set_generator(group, gen, NULL, cofactor);
    ...sometime later....
    EC_GROUP_set_generator(group, gen, order, NULL);

I'm not sure if it makes any sense to code things that way, but theoretically this was possible before and now the first call would fail.

When I wrote the above comment I was actually thinking that the generator could be done that way too, which then makes more sense, i.e. 

    EC_GROUP_set_generator(group, gen, NULL, NULL);
    ...sometime later....
    EC_GROUP_set_generator(group, NULL, order, cofactor);

But I now realise that the generator must always be present.

Comment (mattcaswell) [reply crypto/ec/ec_lib.c]: Ah! No! I'm talking rubbish because of this code:

````
    if (order != NULL) {
        if (!BN_copy(group->order, order))
            return 0;
    } else {
        BN_zero(group->order);
    }

    /* The cofactor is an optional field, so it should be able to be NULL. */
    if (cofactor != NULL) {
        if (!BN_copy(group->cofactor, cofactor))
            return 0;
    } else {
        BN_zero(group->cofactor);
    }
````
i.e. if one or the other of order/cofactor is not present then it is reset to zero

So, I withdraw my objection to this.

Comment (bbbrumley) [reply crypto/ec/ec_lib.c]: Great :+1: 

I will submit a separate PR for `EC_GROUP_set_generator` documentation:

> EC_GROUP_set_generator sets curve parameters that must be agreed by all participants using the curve. These parameters include the generator, the order and the cofactor. The generator is a well defined point on the curve chosen for cryptographic operations. Integers used for point multiplications will be between 0 and n-1 where n is the order. The order multiplied by the cofactor gives the number of points on the curve.

... since without digging in code it's not clear what's allowed. It's like 18yo code so it's not exactly easy to speculate ;)

File: test/ectest.c (hunk 6 above)
Comment (slontis) [reply test/ectest.c]: >do you think it's worth it to add a separate check to test EC_GROUP_check_named_curve()'s behaviour in this corner case?

Its probably not needed.

**Comment by mattcaswell** [comment]
Merged to master.

@bbbrumley - are you able to provide a backport for 1.1.1/1.1.0/1.0.2? As well as the BN_CTX_new_ex issue noted above there seems to be some significant merge conflicts when bringing this back to 1.1.1. I'd really like this to be in the release next week.

**Comment by bbbrumley** [comment]
> are you able to provide a backport

I'll get started on it today.

**Comment by romen** [comment]
#9727 completes this PR w.r.t. `master` with a `CHANGES` entry and a fix to the test logic.

**Comment by romen** [comment]
Merged to `1.1.1-stable` via #9781  .

**Comment by mattcaswell** [comment]
We should create a new PR to add the CHANGES entry into 1.1.1

**Comment by romen** [comment]
> We should create a new PR to add the CHANGES entry into 1.1.1

#9800 appends the mention of CVE-2019-1547 in the 1.1.1 branch.

**Comment by romen** [comment]
For completeness:

- #9799 backports this to `1.0.2-stable`
- #9795 backports this to `1.1.0-stable`
